platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+                });
+            }
+        }
+
+        let mut guard = self.isr.lock(self.smhc);
+        guard.data.op = match (params.kind, params.buffer) {
+            (sdmmc::CommandKind::Control, _) => SmhcOp::Control,
+            (sdmmc::CommandKind::Read(cnt), Some(buf)) => SmhcOp::Read {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            (sdmmc::CommandKind::Write(cnt), Some(buf)) => SmhcOp::Write {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            _ => {
+                tracing::error!("did not provide a buffer for read/write");
+                return Err(sdmmc::Error::Other);
+            }
+        };
+
+        // Statically allocate space for 16 DMA descriptors.
+        // Each descriptor can do a transfer of 4KB, giving a max total transfer of 64KB.
+        // By declaring them in this scope the descriptor memory will live long enough,
+        // however currently this is up to the user to guarantuee.
+        // Safety: a zero'ed descriptor is valid and will simply be ignored by the IDMAC.
+        let mut descriptors: [idmac::Descriptor; 16] = unsafe { core::mem::zeroed() };
+
+        // Perform checks on arguments to make sure we won't overflow the buffer
+        match &guard.data.op {
+            SmhcOp::Read { buf, cnt, .. } | SmhcOp::Write { buf, cnt, .. } => {
+                // Currently we limit the number of data that can be read at once
+                if *cnt > 0x10000 || buf.capacity() < *cnt {
+                    return Err(sdmmc::Error::Data);
+                }
+
+                tracing::debug!("Creating descriptor chain from buffer");
+                let buf_ptr = buf.as_slice().as_ptr();
+                let mut remaining = *cnt;
+                let mut index = 0;
+                while remaining > 0x1000 {
+                    descriptors[index] = idmac::Descriptor::try_from(idmac::DescriptorConfig {
+                        disable_int_on_complete: true,
+                        first: index == 0,
+                        last: false,
+                        buff_size: 0x1000,
+                        buff_addr: unsafe { buf_ptr.add(index * 0x1000) },
+                        next_desc: Some(&descriptors[index + 1] as *const _ as _),
+                    })
+                    .map_err(|_| sdmmc::Error::Other)?;
+                    remaining -= 0x1000;
+                    index += 1;
+                }
+
+                descriptors[index] = idmac::Descriptor::try_from(idmac::DescriptorConfig {
+                    disable_int_on_complete: false,
+                    first: index == 0,
+                    last: true,
+                    buff_size: remaining as u16,
+                    buff_addr: unsafe { buf_ptr.add(index * 0x1000) },
+                    next_desc: None,
+                })
+                .map_err(|_| sdmmc::Error::Other)?;
```

Comment:
take it or leave it, up to you: we could get rid of the final assignment outside of the loop if we changed the loop a little bit:
```suggestion
                while remaining > 0 {
                    let buff_size = core::cmp::min(0x1000, remaining);
                    remaining = remaining.saturating_sub(0x1000);
                    let last = remaining == 0;
                    descriptors[index] = idmac::Descriptor::try_from(idmac::DescriptorConfig {
                        disable_int_on_complete: true,
                        first: index == 0,
                        last,
                        buff_size,
                        buff_addr: unsafe { buf_ptr.add(index * 0x1000) },
                        next_desc: (!last).then_some(|_| (&descriptors[index + 1] as *const _ as _)),
                    })
                    .map_err(|_| sdmmc::Error::Other)?;
                    index += 1;
                }
```
although i'm not sure how much clearer this is, really. up to you...

Reply:
It might make it a bit more complex, but I like it that everything is done inside the loop, so I'm going to apply this suggestion!